### PR TITLE
expand and collapse all

### PR DESCRIPTION
**Pull request by AndyWu2015**
Hey @donat  , I didn't add the test about "double click to open source test file" , cause I find other tests about that, if you need it , I will add it later

### CODE REVIEW

File: org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/BaseExecutionViewTest.groovy [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                true,
+                false,
+                false,
+                false);
+
+        ILaunchConfiguration configuration = CorePlugin.gradleLaunchConfigurationManager().getOrCreateRunConfiguration(attributes)
+        ILaunchConfigurationWorkingCopy launchConfigurationWC = configuration.getWorkingCopy()
+        launchConfigurationWC.setAttribute( "org.eclipse.debug.ui.ATTR_LAUNCH_IN_BACKGROUND", true )
+        launchConfigurationWC.setAttribute( "org.eclipse.debug.ui.ATTR_CAPTURE_IN_CONSOLE", true )
+        launchConfigurationWC.setAttribute( "org.eclipse.debug.ui.ATTR_PRIVATE", true )
```
Comment (donat) [inline org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/BaseExecutionViewTest.groovy]: Can you please explain why these attributes are necessary?

File: org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/ExecutionViewDoubleClickTest.groovy [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        then:
+        !runNode.expanded
+
+        when:
+        runNode.doubleClick()
+
+        then:
+        runNode.expanded
+    }
+
```
Comment (donat) [inline org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/ExecutionViewDoubleClickTest.groovy]: I think we should have an additional test here verifying that double-clicking a test class and method opens the test file. Something like:
```
setup: importProjectWithTest()
when: testNode.doubleClick()
then: assertEditorOpenedOnFile(testFile)
cleanup: closeAllEditors()
```

File: org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/ExecutionViewExpandAndCollapseAllTest.groovy [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import org.eclipse.buildship.ui.test.fixtures.SwtBotSpecification
+import org.eclipse.buildship.ui.util.workbench.WorkbenchUtils
+
+class ExecutionViewExpandAndCollapseAllTest extends BaseExecutionViewTest {
+
+    def "Expand and collapse all"() {
+        setup:
+        def project = sampleProject()
+        importAndWait(project)
+        launchTask('${workspace_loc:/root}', 'foo')
```
Comment (donat) [inline org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/ExecutionViewExpandAndCollapseAllTest.groovy]: The `launchTask()` method should take the root project location from a `File` instance.

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/PluginImages.java [hunk 4]
```diff
@@ -47,8 +47,7 @@
     CANCEL_BUILD_EXECUTION(ImmutableMap.of(ImageState.ENABLED, "icons/full/elcl16/cancel_build_execution.png", ImageState.DISABLED, "icons/full/dlcl16/cancel_build_execution.png")),
     RERUN_BUILD(ImmutableMap.of(ImageState.ENABLED, "icons/full/elcl16/rerun.png", ImageState.DISABLED, "icons/full/dlcl16/rerun.png")),
     RERUN_FAILED_TESTS(ImmutableMap.of(ImageState.ENABLED, "icons/full/elcl16/rerun_failed_tests.png", ImageState.DISABLED, "icons/full/dlcl16/rerun_failed_tests.png")),
-    EXPAND_NODE(ImmutableMap.of(ImageState.ENABLED, "icons/full/elcl16/expand_node.png")),
-    COLLAPSE_NODE(ImmutableMap.of(ImageState.ENABLED, "icons/full/elcl16/collapse_node.png")),
+    EXPAND_ALL(ImmutableMap.of(ImageState.ENABLED, "icons/full/elcl16/expand_all.png")),
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/PluginImages.java]: Out of curiosity, where does this icon come from? I guess only the collapse all icon is available from the shared images.

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/util/nodeselection/ActionShowingContextMenuListener.java [hunk 5]
```diff
@@ -64,6 +69,10 @@ private void handleSelection(IMenuManager manager, NodeSelection selection) {
                 }
             }
         }
+
+        if (isEmpty) {
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/util/nodeselection/ActionShowingContextMenuListener.java]: The `isEmpty` variable is unnecessary. You can replace it with `this.actions.isEmpty()`.

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/NoActionsAvailableAction.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2015 the original author or authors.
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/NoActionsAvailableAction.java]: 2017

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/NoActionsAvailableAction.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/NoActionsAvailableAction.java]: Move the contributor section to the class-level javadoc.

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/NoActionsAvailableAction.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+package org.eclipse.buildship.ui.view;
+
+import org.eclipse.jface.action.Action;
+
+import org.eclipse.buildship.ui.util.nodeselection.NodeSelection;
+import org.eclipse.buildship.ui.util.nodeselection.SelectionSpecificAction;
+
+/**
+ * show disabled "No Actions Available" when there is no available actions for some node
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/NoActionsAvailableAction.java]:
```
/**
 * Action to show in the context menu when no other actions are available.
 *
 * @author Andy Wu <andy.wu@liferay.com> 
 */
```

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/NoActionsAvailableAction.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+ */
+public final class NoActionsAvailableAction extends Action implements SelectionSpecificAction {
+
+    public NoActionsAvailableAction() {
+        setText("No Actions Avaliable");
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/NoActionsAvailableAction.java]: Externalize string, like in the other actions.

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/execution/ExecutionPage.java [hunk 10]
```diff
@@ -313,24 +316,37 @@ private ActionShowingContextMenuListener createContextMenuListener(TreeViewer tr
...
             public void doubleClick(DoubleClickEvent event) {
+                NodeSelection nodeSelection = NodeSelection.from(event.getSelection());
                 OpenTestSourceFileAction openTestSourceFileAction = new OpenTestSourceFileAction(ExecutionPage.this);
-                openTestSourceFileAction.run();
+
+                if (openTestSourceFileAction.isVisibleFor(nodeSelection) && openTestSourceFileAction.isEnabledFor(nodeSelection)) {
+                    openTestSourceFileAction.run();
+                } else if (nodeSelection.isSingleSelection()) {
+                    Object selected = nodeSelection.toList().get(0);
+                    IContentProvider provider = getPageControl().getViewer().getContentProvider();
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/execution/ExecutionPage.java]: For readability, `getPageControl().getViewer()` should be extracted to local variable.

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/PluginImages.java (hunk 4 above)
Comment (AndyWu2015) [reply org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/PluginImages.java]: if you do some search(Ctrl+H) in eclipse, you will find the result view has a expand_all icon , and yeah , it is not in shared images, it is in search related plugins , I just copy it from source.

**Comment by donat** [comment]
Can we help anything regarding this PR?

**Comment by AndyWu2015** [comment]
Sorry I was on my trip last week , I will finish it this week.

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/util/nodeselection/ActionShowingContextMenuListener.java (hunk 5 above)
Comment (AndyWu2015) [reply org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/util/nodeselection/ActionShowingContextMenuListener.java]: I think it is useful , cause "isEmpty" is for checking if no actions are visible for selected node.

File: org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/BaseExecutionViewTest.groovy (hunk 1 above)
Comment (AndyWu2015) [reply org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/BaseExecutionViewTest.groovy]: unnecessary , I just copy from other place , sorry for that.

File: org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/ExecutionViewDoubleClickTest.groovy [hunk 11]
```diff
@@ -0,0 +1,47 @@
+package org.eclipse.buildship.ui.view.execution
+
+import org.eclipse.buildship.ui.test.fixtures.SwtBotSpecification
+import org.eclipse.swtbot.swt.finder.widgets.SWTBotTreeItem
+
+class ExecutionViewDoubleClickTest extends BaseExecutionViewTest {
+
+    def "Double-clicking group nodes expand and collapse children"() {
```
[the remark this answers is not in the document]
Comment (AndyWu2015) [reply org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/ExecutionViewDoubleClickTest.groovy]: done

File: org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/ExecutionViewDoubleClickTest.groovy [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+        !root.expanded
+
+        when:
+        root.doubleClick()
+
+        then:
+        root.expanded
+
+        when:
+        SWTBotTreeItem runNode =  root.getNode('Run tasks')
```
[the remark this answers is not in the document]
Comment (AndyWu2015) [reply org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/ExecutionViewDoubleClickTest.groovy]: done

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/NoActionsAvailableAction.java (hunk 6 above)
Comment (AndyWu2015) [reply org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/NoActionsAvailableAction.java]: done

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/NoActionsAvailableAction.java (hunk 7 above)
Comment (AndyWu2015) [reply org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/NoActionsAvailableAction.java]: done

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/NoActionsAvailableAction.java (hunk 9 above)
Comment (AndyWu2015) [reply org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/NoActionsAvailableAction.java]: done

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/execution/ExecutionPage.java (hunk 10 above)
Comment (AndyWu2015) [reply org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/execution/ExecutionPage.java]: done

File: org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/ExecutionViewExpandAndCollapseAllTest.groovy (hunk 3 above)
Comment (AndyWu2015) [reply org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/ExecutionViewExpandAndCollapseAllTest.groovy]: done

File: org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/BaseExecutionViewTest.groovy [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    protected void removeCloseableGradleConsoles() {
+        IConsoleManager consoleManager = ConsolePlugin.default.consoleManager
+        List<GradleConsole> gradleConsoles = consoleManager.consoles.findAll { console -> console instanceof GradleConsole && console.isCloseable() }
+        consoleManager.removeConsoles(gradleConsoles as IConsole[])
+    }
+
+    protected File sampleProject() {
+        dir('root') { file 'build.gradle', """
```
[the remark this answers is not in the document]
Comment (AndyWu2015) [reply org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/BaseExecutionViewTest.groovy]: Ok,  I will move it to children classes separately.

File: org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/ExecutionViewDoubleClickTest.groovy (hunk 2 above)
Comment (AndyWu2015) [reply org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/ExecutionViewDoubleClickTest.groovy]: done

**Comment by AndyWu2015** [comment]
Hey donat , I've finished most the change requests and some change request I didn't finish I leave some description , but I met an issue for your help:
when I run the execution view tests I wrote, it sometimes fail cause can't generate all of the execution tree , I debugged and found this 
tree-error

do you have any idea about this ?
thanks

**Comment by donat** [comment]
Thanks for all the changes, I'll have a look shortly. In the meanwhile, I checked the problem with the execution tree not being complete. I was able to reproduce the issue. You don't have to worry about it - unless you really want to :) - I'll fix it later.

**Comment by donat** [comment]
The execution view problem is tracked at #586.

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/util/nodeselection/ActionShowingContextMenuListener.java (hunk 5 above)
Comment (donat) [reply org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/util/nodeselection/ActionShowingContextMenuListener.java]: My bad, I thought that the `isEmpty = true` statement is outside of `if (action.isVisibleFor(selection)) {}`

**Comment by donat** [comment]
@AndyWu2015 FYI: I did a few small adjustments before merging: https://github.com/eclipse/buildship/commit/2b56b0a976ef3e0803fc319ff73574126de3ed24

**Comment by AndyWu2015** [comment]
Thanks, @donat  hope I can do more for buildship :)

**Pull request merged**